index.js

```diff
@@ -873,6 +879,38 @@ class KeyringController extends EventEmitter {
       );
     }
   }
+
+  async _newKeyring(type, data) {
+    const keyringBuilder = this.getKeyringBuilderForType(type);
+
+    if (!keyringBuilder) {
+      return undefined;
+    }
+
+    const keyring = keyringBuilder();
+
+    await keyring.deserialize(data);
+
+    if (keyring.init) {
+      await keyring.init();
+    }
+
+    return keyring;
+  }
 }
 
-module.exports = KeyringController;
+function keyringBuilderFactory(KeyringClass, BridgeClass) {
+  const builder = () => {
+    const constructorDependencies = BridgeClass ? new BridgeClass() : undefined;
```

Comment:
I'm torn on whether this is a good idea. I see the advantage of including a "builder" function here directly, but maybe it would be better to encourage using custom builder functions instead.

Most bridges will require initialization, for example, but this offers no way to do that.

Maybe we could instead offer a simpler "builder" function just for the simple keyrings, and recommend a custom builder function for keyrings using bridges?

Reply:
I was at odds with doing this as well, probably best to leave a simple one as you suggest though.

What tilted me towards creating this method was the odd situation in which we have to make sure that the builder function returned must have a `type` property, which is not something particularly easy to document.

I've replaced it with the simpler version, so the one with the bridge will probably have to be added to `metamask-controller` in order to instantiate some of the keyrings (the QR one probably won't need a bridge and can use this function too).